server/coordinator/procedure/create_partition_table.go

```diff
@@ -0,0 +1,342 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package procedure
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/CeresDB/ceresdbproto/golang/pkg/metaservicepb"
+	"github.com/CeresDB/ceresmeta/pkg/log"
+	"github.com/CeresDB/ceresmeta/server/cluster"
+	"github.com/CeresDB/ceresmeta/server/coordinator/eventdispatch"
+	"github.com/CeresDB/ceresmeta/server/storage"
+	"github.com/looplab/fsm"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	eventCreatePartitionTable = "EventCreatePartitionTable"
+	eventCreateDataTables     = "EventCreateDataTables"
+	eventOpenPartitionTables  = "EventOpenPartitionTables"
+	eventFinish               = "EventSuccess"
+
+	stateBegin                = "StateCreatePartitionTableBegin"
+	stateCreatePartitionTable = "StateCreatePartitionTable"
+	stateCreateDataTables     = "StateCreatePartitionTableCreateDataTables"
+	stateOpenPartitionTables  = "StateOpenPartitionTables"
+	stateFinish               = "StateCreatePartitionTableFinish"
+)
+
+var (
+	createPartitionTableEvents = fsm.Events{
+		{Name: eventCreatePartitionTable, Src: []string{stateBegin}, Dst: stateCreatePartitionTable},
+		{Name: eventCreateDataTables, Src: []string{stateCreatePartitionTable}, Dst: stateCreateDataTables},
+		{Name: eventOpenPartitionTables, Src: []string{stateCreateDataTables}, Dst: stateOpenPartitionTables},
+		{Name: eventFinish, Src: []string{stateOpenPartitionTables}, Dst: stateFinish},
+	}
+	createPartitionTableCallbacks = fsm.Callbacks{
+		eventCreatePartitionTable: createPartitionTableCallback,
+		eventCreateDataTables:     createDataTablesCallback,
+		eventOpenPartitionTables:  openPartitionTablesCallback,
+		eventFinish:               finishCallback,
+	}
+)
+
+type CreatePartitionTableProcedure struct {
+	id       uint64
+	fsm      *fsm.FSM
+	cluster  *cluster.Cluster
+	dispatch eventdispatch.Dispatch
+	storage  Storage
+
+	req *metaservicepb.CreateTableRequest
+
+	partitionTableShards []cluster.ShardNodeWithVersion
+	dataTablesShards     []cluster.ShardNodeWithVersion
+
+	onSucceeded func(cluster.CreateTableResult) error
+	onFailed    func(error) error
+
+	lock  sync.RWMutex
+	state State
+}
+
+func NewCreatePartitionTableProcedure(id uint64, cluster *cluster.Cluster, dispatch eventdispatch.Dispatch, storage Storage, req *metaservicepb.CreateTableRequest, partitionTableShards []cluster.ShardNodeWithVersion, dataTablesShards []cluster.ShardNodeWithVersion, onSucceeded func(cluster.CreateTableResult) error, onFailed func(error) error) *CreatePartitionTableProcedure {
```

Comment:
The input params are too many, and maybe we should define struct to replace most of these parameters.